### PR TITLE
utils: memory order as parameter for get_linear_index

### PR DESCRIPTION
**Pull request by fweik**
This is mostly to document what the memory order is, and I need
row major for the p3m. @jngrad do you have an idea how to document
the tag parameter in doxygen? I'd rather not name it because it's
not used...

### CODE REVIEW

**Comment by jngrad** [comment]
@fweik unnamed parameters cannot be documented via `@param`. You can either mention in the long description how the last argument is just a tag for the function template, or name the argument, e.g. `major_order`. In the future with C++17 these 3 functions could probably be condensed into a single function template using `if(std::is_same<T, MemoryOrder::RowMajor>::value)`.

File: src/utils/include/utils/index.hpp [hunk 1]
```diff
@@ -58,28 +58,51 @@ inline void unravel_index(InputIterator dimensions_begin,
   }
 }
 
+namespace MemoryOrder {
+constexpr struct RowMajor {
+} row_major;
+constexpr struct ColumnMajor { } coulumn_major; } // namespace MemoryOrder
```
Comment (jngrad) [inline src/utils/include/utils/index.hpp]: typo + formatting
```suggestion
constexpr struct ColumnMajor {
} column_major;
} // namespace MemoryOrder
```

Comment (fweik) [reply src/utils/include/utils/index.hpp]: Actually this is what clang-format wants

File: src/utils/include/utils/index.hpp [hunk 2]
```diff
@@ -58,28 +58,51 @@ inline void unravel_index(InputIterator dimensions_begin,
...
+
 /*************************************************************/
 /** \name Three dimensional grid operations                  */
 /*************************************************************/
 /*@{*/
 
 /** get the linear index from the position (@p a,@p b,@p c) in a 3D grid
- *  of dimensions @p adim.
+ *  of dimensions @p adim, usiung column major memory order.
```
Comment (jngrad) [inline src/utils/include/utils/index.hpp]:
```suggestion
 *  of dimensions @p adim, using column major memory order.
```

**Comment by fweik** [comment]
Well there is no real reason here for static dispatch, see updated version.

**Review by RudolfWeeber — APPROVED** [review]

**Comment by jngrad** [comment]
But the PR didn't pass CI... We still need to replace the switch statement by an if/else.

**Comment by RudolfWeeber** [comment]
Sorry, didn notice that.

**Comment by RudolfWeeber** [comment]
bors cancel

**Comment by fweik** [comment]
I like the switch better

**Review by jngrad — APPROVED** [review]

**Pull request merged**